util/network-devp2p/src/enr.rs

```
@@ -0,0 +1,437 @@
+        self.pairs.insert(value.enr_key(), value);
+    }
+
+    fn invalidate(&mut self) {
+        if self.signature.is_some() {
+            self.seq += 1;
+        }
+
+        self.signature = None;
+        self.raw = None;
+    }
+
+    pub fn signature(&self) -> Option<Vec<u8>> {
+        self.signature.clone()
+    }
+
+    /// Get the scheme represented by a string.
+    /// Since parameter `id` of `get()` has a `Id` type,
+    /// `get` can call `enr_key()`.
+    /// So, we just pass it with empty string.
+    pub fn scheme(&self) -> Option<String> {
+        let id = Id("".to_owned());
+        let value = self.get(&id);
+        match value {
+            Some(Value::Id(id)) => Some(id.0.clone()),
+            _ => {
+                error!(target: "network", "Can't get scheme for ENR");
+                None
+            },
+        }
+    }
+
+    pub fn set_signature(&mut self, scheme: &dyn Scheme, sig: &Vec<u8>) {
+        if !scheme.verify(self, &sig) {
+            error!(target: "network", "Can't verify scheme");
+            return
+        }
+
+        self.signature = Some(sig.clone());
+        let raw = encode(self);
+        self.raw = Some(raw);
+    }
+
+    pub fn reset_signature(&mut self) {
+        self.signature = None;
+        self.raw = None;
+    }
+
+    pub fn set_seq(&mut self, seq: u64) {
+        self.seq = seq;
+        self.signature = None;
+        self.raw = None;
+    }
+
+    pub fn get_seq(&self) -> u64 {
+        self.seq
+    }
+}
+
+impl Encodable for Record {
+    fn rlp_append(&self, s: &mut RlpStream) {
+        s.begin_list(self.pairs.len() * 2 + 3)
+            .append(&self.signature)
+            .append(&self.seq)
+            .append(&self.raw);
+
+        for (key, value) in &self.pairs {
+            s.append(key).append(value);
+        }
+
+        if s.len() > ENR_MAX_SIZE {
+            error!(target: "network", "Encoded bytes exceed limit(300 bytes), length is {:?}", s.len())
```

we should have a way to return an error instead of just printing a warning